File: 0100-same-tree/0100-same-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if (!p && !q) return true;
        if ((p && !q) || (!p && q)) return false;
        
        queue<pair<TreeNode*, TreeNode*>> queue;
        queue.push(make_pair(p, q));
        
        while (!queue.empty()){
            auto[p_derive, q_derive] = queue.front();
            queue.pop();
            
            if (p_derive -> val != q_derive -> val) return false;
            
            if (p_derive -> left && q_derive -> left) queue.push(make_pair(p_derive -> left, q_derive -> left));
            else if ((!p_derive -> left && q_derive -> left) || (p_derive -> left && !q_derive -> left)) return false;
            
            if (p_derive -> right && q_derive -> right) queue.push(make_pair(p_derive -> right, q_derive -> right));
            else if ((!p_derive -> right && q_derive -> right) || (p_derive -> right && !q_derive -> right)) return false;
        }
        
        return true;
    }
};

File: 0100-same-tree/0100-same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        
        if not p and not q:
            return True
        if (p and not q) or (not p and q):
            return False
        
        queue = deque([(p, q)])
        
        while queue:
            p_derive, q_derive = queue.popleft()
            if p_derive.val != q_derive.val:
                return False
            
            if p_derive.left and q_derive.left:
                queue.append((p_derive.left, q_derive.left))
            elif (not p_derive.left and q_derive.left) or (p_derive.left and not q_derive.left):
                return False
            
            if p_derive.right and q_derive.right:
                queue.append((p_derive.right, q_derive.right))
            elif (not p_derive.right and q_derive.right) or (p_derive.right and not q_derive.right):
                return False
            
        return True